groups/bsl/bslstl/bslstl_inplace.h

```diff
@@ -0,0 +1,74 @@
+// bslstl_inplace.h                                                   -*-C++-*-
+
+#ifndef INCLUDED_BSLSTL_INPLACE
+#define INCLUDED_BSLSTL_INPLACE
+
+#include <bsls_ident.h>
+BSLS_IDENT("$Id: $")
+
+//@PURPOSE: Provide a standard-compliant in place construction tag types.
+//
+//@CLASSES:
+//  bsl::in_place_t: tag type for in-place construction
+//
+//@DESCRIPTION: This component provides an implementation of a standard
+// compliant tag type for in-place construction, 'bsl::in_place_t'.  This tag
+// type is used in constructors of bsl::optional to indicate that the contained
+// object should be constructed in-place.  Note that the standard currently has
+// two other in-place construction tag types, 'std::in_place_type_t' and
+// 'std::in_place_index_t'. There is currently no need to provide the
+// equivalent tag types in bsl, but if such need arises, the name of this
+// header has been chosen for the purpose of being the designated place for all
+// in-place construction tags.
+
+#include <bslscm_version.h>
+
+#include <bsls_compilerfeatures.h>
+#include <bsls_keyword.h>
+
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+#include <utility> // for std::in_place_t
+#endif  // BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+
+namespace bsl {
+
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+
+using in_place_t = std::in_place_t;
+using std::in_place;
+
+#else
+
+                              // ================
+                              // class in_place_t
+                              // ================
+struct in_place_t {
+    // This trivial tag type is passed to the constructors of types that
+    // contain a single object to indicate that the contained object should be
+    // constructed in-place.
+};
+
+static const BSLS_KEYWORD_CONSTEXPR in_place_t in_place = in_place_t();
+// Value of type 'in_place_t' used as an argument to functions that take an
+// 'in_place_t' argument.
```

Comment:
Sometimes bde-format isn't doing the best job (which is very unfortunate).  I think this unfortunately requires some manual fiddling (to add an indent).  To fix such cases I often do something like adding a fake function declaration, then formatting with bde-format (it will indent) and then removing the fake function declaration.  Such annoying things are helped a lot by the integration with editor.  I suggest that probably the easiest way is to use VS Code for formatting stuff - the integration is very easy.

Reply:
what should be indented here and by how much ? As in, what rule isn't followed ? I'm assuming it's the documentation of the static variable, but I see other places where there is no indentation and I wonder if those need to be fixed aswell (bslstl_optional.h lines 163 and 210, for example). I'll fix it, I just need to know the rule for future :)

Reply:
We don't have many of such globals, but we do have them - e.g., in `bdlb_nullopt.h`.  I'd say this falls into category of variable documentation (btw, do you have access to our coding standards?) and in this case, I'd just indent the doc (our usual 4 spaces).  Other places that you mentioned also need to be fixed.

Reply:
Also, if you notice in `bdlb_nullopt`, we have a different thing for CPP03 - I assume this is to avoid a static object in every translation unit that includes it.  Was that ever discussed for this `optional` work?

Reply:
i "discussed" it bde_verify tool :) Originally, I made all of these extern, but bde_verify wasn't happy with that, so I checked how allocator_arg is done hence I made them static. I assumed that's the done way. I have no strong feelings either way. I'll indent and swap over to bdlb_nullopt.h approach.
I need to check my inbox, there may have been an email with a link to the coding standard a long time ago. If you have it at hand, I wouldn't mind having it sent to me again.

Reply:
I see :) In addition, Alisdair suggested that we have no real reason to alias std::nullopt(_t)/inplace(_t) when they are available.  Not doing so will (1) slightly reduce the code size and (2) avoid potential issues with increasing the number of associated namespaces for ADL and CTAD.

Reply:
I don't think i quite understand taht comment. 
C++14 implementation of bsl::optional for non AA types uses std::optional implementation. This means that the tag types the user uses (i.e. bsl ones) are picked up by the std::optional. We either have bsl tag types alias to std tag types or we provide a conversion mechanism.
I would personally find the conversion solution "awkward"